### PR TITLE
feat: added notify route functionality (MAPCO-4709)

### PR DESCRIPTION
**Pull request by nitzanmorr**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

Further  information:

### CODE REVIEW

File: tests/unit/tasks/models/tasksModel.spec.ts [hunk 1]
```diff
@@ -1,10 +1,32 @@
...
+import nock from 'nock';
+import { TaskHandler as QueueClient } from '@map-colonies/mc-priority-queue';
 import { TasksManager } from '../../../../src/tasks/models/tasksManager';
-
-let tasksManager: TasksManager;
+import { configMock, registerDefaultConfig, clear as clearConfig } from '../../../mocks/configMock';
+import { IJobManagerConfig } from '../../../../src/common/interfaces';
 
 describe('TasksManager', () => {
+  let jobManagerClient: TasksManager;
```
Comment (CL-SHLOMIKONCHA) [inline tests/unit/tasks/models/tasksModel.spec.ts:9]: taskManager instead of 'jobManagerClient'

**Review by CL-SHLOMIKONCHA — CHANGES_REQUESTED** [review]

File: config/custom-environment-variables.json [hunk 2]
```diff
@@ -38,5 +38,21 @@
         }
       }
     }
+  },
+  "jobManagement": {
+    "config": {
+      "jobManagerBaseUrl": "JOB_MANAGER_URL",
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply config/custom-environment-variables.json:44]: changed

File: config/custom-environment-variables.json [hunk 3]
```diff
@@ -38,5 +38,21 @@
...
+        "baseUrl": "HEARTBEAT_BASE_URL",
+        "intervalMs": "HEARTBEAT_INTERVAL_MS"
+      },
+      "dequeueIntervalMs": "JOB_MANAGER_DEQUEUE_INTERVAL_MS"
+    }
+  },
+  "taskTypes": {
+    "init": "TASK_TYPES_INIT",
+    "tilesMerging": "TASK_TYPES_TILES_MERGING",
+    "polygonParts": "TASK_TYPES_POLYGON-PARTS",
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply config/custom-environment-variables.json:55]: accepted

File: config/default.json [hunk 4]
```diff
@@ -23,6 +23,28 @@
         "enabled": true,
         "options": null
       }
+    },
+    "httpRetry": {
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply config/default.json:27]: fixed

File: config/default.json [hunk 5]
```diff
@@ -23,6 +23,28 @@
...
+      "attempts": 5,
+      "delay": "exponential",
+      "shouldResetTimeout": true,
+      "disableHttpClientLogs": true
+    }
+  },
+  "jobManagement": {
+    "config": {
+      "jobManagerBaseUrl": "https://common-job-manager-route-raster-dev.apps.j1lk3njp.eastus.aroapp.io/",
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply config/default.json:36]: changed

File: src/common/errors.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+export class IrrelevantOperationStatusError extends Error {
+  public constructor(message?: string) {
+    super(message);
+    this.name = 'IrrelevantOperationStatusError';
+  }
+}
+
+export class TasksNotFoundError extends Error {
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/common/errors.ts:8]: changed to the error from `@map-colonies/error-types`

File: src/tasks/controllers/tasksController.ts [hunk 7]
```diff
@@ -3,25 +3,35 @@ import { Meter } from '@opentelemetry/api-metrics';
...
-}
+import { TaskNotificationRequest } from '../../common/interfaces';
+import { IrrelevantOperationStatusError, TasksNotFoundError } from '../../common/errors';
 
 type TaskNotificationHandler = RequestHandler<TaskNotificationRequest, undefined, undefined>;
 
 @injectable()
 export class TasksController {
   public constructor(
     @inject(SERVICES.LOGGER) private readonly logger: Logger,
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/controllers/tasksController.ts:18]: removed

File: src/tasks/controllers/tasksController.ts [hunk 8]
```diff
@@ -3,25 +3,35 @@ import { Meter } from '@opentelemetry/api-metrics';
...
-    @inject(TasksManager) private readonly manager: TasksManager,
-    @inject(SERVICES.METER) private readonly meter: Meter
+    @inject(SERVICES.METER) private readonly meter: Meter,
+    @inject(TasksManager) private readonly manager: TasksManager
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/controllers/tasksController.ts:20]: changed

File: src/tasks/models/tasksManager.ts [hunk 9]
```diff
@@ -1,8 +1,114 @@
...
+
+  public constructor(
+    @inject(SERVICES.LOGGER) private readonly logger: Logger,
+    @inject(SERVICES.QUEUE_CLIENT) private readonly queueClient: QueueClient
+  ) {
+    this.jobManager = this.queueClient.jobManagerClient;
+    this.taskTypes = config.get<ITaskTypesConfig>('taskTypes');
+  }
+
+  public async handleTaskNotification(taskId: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:31]: added

File: src/tasks/models/tasksManager.ts [hunk 10]
```diff
@@ -1,8 +1,114 @@
...
+    const task = await this.findTask({ id: taskId });
+    if (!task) {
+      throw new TasksNotFoundError(`Task ${taskId} not found`);
+    }
+    if (task.status === OperationStatus.FAILED) {
+      await this.failJob(task.jobId);
+      this.logger.info({ msg: `Failed job: ${task.jobId}` });
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:38]: moved

File: src/tasks/models/tasksManager.ts [hunk 11]
```diff
@@ -1,8 +1,114 @@
...
+    } else {
+      throw new IrrelevantOperationStatusError(`Expected to get a 'Completed' or 'Failed' task' but instead got '${task.status}'`);
+    }
+  }
+
+  public async handleCompletedTask(job: IJobResponse<unknown, unknown>, task: ITaskResponse<unknown>): Promise<void> {
+    const initTask = await this.findTask({ jobId: job.id, type: this.taskTypes.init });
+
+    if (!initTask) {
+      this.logger.debug({ msg: 'Did nothing because init task was not found' });
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:51]: changed message

File: src/tasks/models/tasksManager.ts [hunk 12]
```diff
@@ -1,8 +1,114 @@
...
+      return;
+    }
+
+    if (initTask.status !== OperationStatus.COMPLETED) {
+      this.logger.debug({ msg: 'Did nothing because init task is not completed' });
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:56]: changed message

File: src/tasks/models/tasksManager.ts [hunk 13]
```diff
@@ -1,8 +1,114 @@
...
+      return;
+    }
+
+    const taskHasSubsequentTask = task.type === this.taskTypes.tilesMerging || task.type === this.taskTypes.polygonParts;
+
+    if (job.completedTasks === job.taskCount && taskHasSubsequentTask) {
+      await this.createNextTask(task.type, job);
+    } else {
+      this.logger.debug({ msg: 'Skipping task creation' });
+      const updatedPercentage = calculateTaskPercentage(job.completedTasks, job.taskCount);
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:66]: moved the function inside the updateJobPercentage argument

File: src/tasks/models/tasksManager.ts [hunk 14]
```diff
@@ -1,8 +1,114 @@
...
+    const task = await this.jobManager.findTasks(body);
+    return task?.[0];
+  }
+
+  public async updateJobPercentage(jobId: string, desiredPercentage: number): Promise<void> {
+    await this.jobManager.updateJob(jobId, { percentage: desiredPercentage });
+    this.logger.info({ msg: `Updated percentages for job: ${jobId}` });
+  }
+
+  public async createNextTask(previousTaskType: string, job: IJobResponse<unknown, unknown>): Promise<void> {
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:91]: changed

File: src/tasks/models/tasksManager.ts [hunk 15]
```diff
@@ -1,8 +1,114 @@
...
+    let nextTaskType: string;
+    switch (previousTaskType) {
+      case this.taskTypes.tilesMerging: {
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:94]: removed it

File: src/tasks/models/tasksManager.ts [hunk 16]
```diff
@@ -1,8 +1,114 @@
...
+      }
+      default: {
+        throw new InvalidArgumentError(`No subsequent task is defined for task type '${previousTaskType}'`);
+      }
+    }
+
+    const existingTask = await this.jobManager.findTasks({ jobId: job.id, type: nextTaskType });
+    if (!existingTask) {
+      await this.createTask(job.id, nextTaskType);
+      const updatedPercentage = calculateTaskPercentage(job.completedTasks, job.taskCount + 1);
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:110]: moved it inside `updateJobPercentage()`

File: tests/unit/tasks/models/tasksModel.spec.ts (hunk 1 above)
Comment (nitzanmorr) [reply tests/unit/tasks/models/tasksModel.spec.ts:9]: changed it

File: src/tasks/models/tasksManager.ts [hunk 17]
```diff
@@ -1,8 +1,114 @@
...
+      this.logger.debug({ msg: 'Did nothing because init task is not completed' });
+      return;
+    }
+
+    const taskHasSubsequentTask = task.type === this.taskTypes.tilesMerging || task.type === this.taskTypes.polygonParts;
+
+    if (job.completedTasks === job.taskCount && taskHasSubsequentTask) {
+      await this.createNextTask(task.type, job);
+    } else {
+      this.logger.debug({ msg: 'Skipping task creation' });
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts:65]: changed log to "Updating job percentage; no need for task creation"

**Review by CL-SHLOMIKONCHA — CHANGES_REQUESTED** [review]

File: src/tasks/models/tasksManager.ts [hunk 18]
```diff
@@ -62,18 +64,24 @@ export class TasksManager {
     if (job.completedTasks === job.taskCount && taskHasSubsequentTask) {
       await this.createNextTask(task.type, job);
     } else {
-      this.logger.debug({ msg: 'Skipping task creation' });
-      const updatedPercentage = calculateTaskPercentage(job.completedTasks, job.taskCount);
-      await this.updateJobPercentage(job.id, updatedPercentage);
+      this.logger.debug({ msg: 'Updating job percentage; no need for task creation' });
+      await this.updateJobPercentage(job.id, calculateTaskPercentage(job.completedTasks, job.taskCount));
```
[the remark this answers is not in the document]
Comment (nitzanmorr) [reply src/tasks/models/tasksManager.ts]: reverted

**Review by CL-SHLOMIKONCHA — APPROVED** [review]

**Comment by CL-SHLOMIKONCHA** [comment]
test will be fixed next PR

**Pull request merged**